src/Expander.hs

```diff
@@ -926,28 +926,58 @@ runTask (tid, localData, task) = withLocal localData $ do
             addConstructor cn dt ctor ts
           linkDeclOutputScopes outScopesDest scs
     AwaitingPattern patPtr ty dest stx -> do
-      ready <-
-        case patPtr of
-          Left pptr -> isJust . view (expanderCompletedPatterns . at pptr) <$> getState
-          Right tptr -> isJust . view (expanderCompletedTypePatterns . at tptr) <$> getState
-      if (not ready)
+      ready <- do
+        let patternComplete ptr =
+              (view (expanderCompletedPatterns . at ptr) <$> getState) >>=
+              \case
+                Nothing ->
+                  pure False
+                Just layer ->
+                  and <$> traverse patternComplete layer
+        patternComplete patPtr
+        -- TODO traverse the tree and get the layers out, collecting the set of pointers to later accumulate the binders
```

Comment:
I don't quite understand this TODO. Are you saying that there's some piece of functionality which is missing from this PR, or that one day it would be nice to get rid of `expanderPatternBinders` and to instead put that information in `expanderCompletedPatterns`?

Reply:
It was to write the code immediately below it :-)